frontend/src/metabase/common/components/QuestionPicker/components/QuestionPicker.tsx

```diff
@@ -0,0 +1,180 @@
+import { getPathLevelForItem } from "../../CollectionPicker/utils";
+import {
+  LoadingSpinner,
+  NestedItemPicker,
+  type PickerState,
+} from "../../EntityPicker";
+import type { QuestionPickerOptions, QuestionPickerItem } from "../types";
+import {
+  generateKey,
+  getCollectionIdPath,
+  getStateFromIdPath,
+  isFolder,
+} from "../utils";
+
+export const defaultOptions: QuestionPickerOptions = {
+  showPersonalCollections: true,
+  showRootCollection: true,
+  allowCreateNew: false,
+};
+interface QuestionPickerProps {
+  onItemSelect: (item: QuestionPickerItem) => void;
+  initialValue?: Pick<QuestionPickerItem, "model" | "id">;
+  options: QuestionPickerOptions;
+  models?: SearchModel[];
+}
+
+const useGetInitialCollection = (
+  initialValue?: Pick<QuestionPickerItem, "model" | "id">,
+) => {
+  const isQuestion =
+    initialValue && ["card", "dataset"].includes(initialValue.model);
+
+  const cardId = isQuestion ? Number(initialValue.id) : undefined;
+
+  const { data: currentQuestion, error: questionError } = useQuestionQuery({
+    id: cardId,
+    enabled: !!cardId,
+  });
+
+  const collectionId =
+    isQuestion && currentQuestion
+      ? currentQuestion?.collectionId()
+      : initialValue?.id;
+
+  const { data: currentCollection, error: collectionError } =
+    useCollectionQuery({
+      id: (isValidCollectionId(collectionId) && collectionId) || "root",
+      enabled: !isQuestion || !!currentQuestion,
+    });
+
+  return {
+    currentQuestion: currentQuestion?._card,
+    currentCollection,
+    isLoading: !currentCollection,
+    error: questionError ?? collectionError,
+  };
+};
+
+export const QuestionPicker = ({
+  onItemSelect,
+  initialValue,
+  options,
+  models = ["dataset", "card"],
+}: QuestionPickerProps) => {
+  const [path, setPath] = useState<
+    PickerState<QuestionPickerItem, SearchRequest>
+  >(() =>
+    getStateFromIdPath({
+      idPath: ["root"],
+      models,
+    }),
+  );
+
+  const {
+    currentCollection,
+    currentQuestion,
+    error,
+    isLoading: loadingCurrentCollection,
+  } = useGetInitialCollection(initialValue);
+
+  const userPersonalCollectionId = useSelector(getUserPersonalCollectionId);
+
+  const onFolderSelect = useCallback(
+    ({ folder }: { folder: QuestionPickerItem }) => {
+      const newPath = getStateFromIdPath({
+        idPath: getCollectionIdPath(folder, userPersonalCollectionId),
+        models,
+      });
+      setPath(newPath);
+      onItemSelect(folder);
+    },
+    [setPath, onItemSelect, userPersonalCollectionId, models],
+  );
+
+  const handleItemSelect = useCallback(
+    (item: QuestionPickerItem) => {
+      // set selected item at the correct level
+      const pathLevel = getPathLevelForItem(
+        item,
+        path,
+        userPersonalCollectionId,
+      );
+
+      const newPath = path.slice(0, pathLevel + 1);
+      newPath[newPath.length - 1].selectedItem = item;
```

Comment:
Alas, a similar line in `handleItemSelect` did throw a TypeError for me. Here's a [loom](https://www.loom.com/share/1019019f93ad4776bfdaa0e0cf8c3766?sid=e3da2567-209a-4972-8033-b3d81c069717)

Reply:
Stack trace:

```
Uncaught TypeError: Cannot set properties of undefined (setting 'selectedItem')
    at QuestionPicker.tsx:117:1
    at handleClick (NestedItemPicker.tsx:56:1)
    at onClick (NestedItemPicker.tsx:86:1)
    at onClick (ItemList.tsx:85:1)
    at handleClick (NavLink.js:101:1)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945:1)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994:1)
    at invokeGuardedCallback (react-dom.development.js:4056:1)
    at invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:4070:1)
    at executeDispatch (react-dom.development.js:8243:1)
```

Reply:
yikes! good catch. 
1. there was a bug with top level collections being treated inconsitently: sometimes as folders, other times as items. this has been fixed.
2. i've also modified the path finding function to return zero if it ever can't find the correct level. this shouldn't ever happen, but if it does, we'll return you to the top level instead of throwing an error.
3.